cpp/include/cudf/table/row_operator3.cuh

```diff
@@ -0,0 +1,315 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/detail/utilities/assert.cuh>
+#include <cudf/detail/utilities/hash_functions.cuh>
+#include <cudf/sorting.hpp>
+#include <cudf/table/row_operators.cuh>
+#include <cudf/table/table_device_view.cuh>
+#include <cudf/utilities/traits.hpp>
+#include <cudf/utilities/type_dispatcher.hpp>
+
+#include <thrust/equal.h>
+#include <thrust/swap.h>
+#include <thrust/transform_reduce.h>
+
+#include <limits>
+#include <memory>
+#include <utility>
+
+namespace cudf {
+namespace experimental {
+
+template <cudf::type_id t>
+struct non_nested_id_to_type {
+  using type = std::conditional_t<cudf::is_nested(data_type(t)), void, id_to_type<t>>;
+};
+
+namespace lexicographic_comparison {
+
+/**
+ * @brief Performs a relational comparison between two elements in two columns.
+ *
+ * @tparam Nullate A cudf::nullate type describing how to check for nulls.
+ */
+template <typename Nullate>
+class element_relational_comparator {
+ public:
+  /**
+   * @brief Construct type-dispatched function object for performing a
+   * relational comparison between two elements.
+   *
+   * @note `lhs` and `rhs` may be the same.
+   *
+   * @param lhs The column containing the first element
+   * @param rhs The column containing the second element (may be the same as lhs)
+   * @param has_nulls Indicates if either input column contains nulls.
+   * @param null_precedence Indicates how null values are ordered with other values
+   */
+  __host__ __device__ element_relational_comparator(Nullate has_nulls,
+                                                    column_device_view lhs,
+                                                    column_device_view rhs,
+                                                    null_order null_precedence,
+                                                    int depth = std::numeric_limits<int>::max())
+    : lhs{lhs}, rhs{rhs}, nulls{has_nulls}, null_precedence{null_precedence}, depth{depth}
+  {
+  }
+
+  __host__ __device__ element_relational_comparator(Nullate has_nulls,
+                                                    column_device_view lhs,
+                                                    column_device_view rhs)
+    : lhs{lhs}, rhs{rhs}, nulls{has_nulls}
+  {
+  }
+
+  /**
+   * @brief Performs a relational comparison between the specified elements
+   *
+   * @param lhs_element_index The index of the first element
+   * @param rhs_element_index The index of the second element
+   * @return Indicates the relationship between the elements in
+   * the `lhs` and `rhs` columns.
+   */
+  template <typename Element,
+            std::enable_if_t<cudf::is_relationally_comparable<Element, Element>()>* = nullptr>
+  __device__ thrust::pair<weak_ordering, int> operator()(size_type lhs_element_index,
```

Comment:
Let's use `cuda::std::pair` since `thrust::pair` isn't long for this world.

Reply:
Or even a custom type since this is used quite a bit.

Reply:
I changed it to `cuda::std::pair` and surprisingly, the performance deteriorated by 10-20% across all row sizes. Custom pod struct also had the exact same performance as `cuda::std::pair`. Keeping `cuda::std::pair` but it's interesting to note that thrust::pair is a tiny bit faster.

Reply:
Try making sure the custom struct is naturally aligned, i.e., `sizeof(T) == alignof(T)`.

Reply:
Well, the custom struct with alignment had absolutely no effect. Confirmed `sizeof(T) == alignof(T)` with a static assert